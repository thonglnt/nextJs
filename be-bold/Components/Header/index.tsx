// Mock
import { MockDataMenu } from "@/mock/mockDataMenu";
import Link from "next/link";
import { useState } from "react";
import Sidebar from "../Sidebar";

const Header = () => {
  const [isOpen, setIsOpen] = useState(false);

  const renderDataMenu = () =>
    MockDataMenu.map((menuItem) => (
      <li key={menuItem} className="nav-item menu">
        <div className="wrapper-hr-tag">
          <hr className="hr-tag" />
        </div>
        <a className="nav-link" aria-current="page" href="#">
          {menuItem}
        </a>
      </li>
    ));

  const handleOpenCard = () => {
    setIsOpen(!isOpen);
  };

  return (
    <div className="container-fluid d-flex justify-content-between mt-3 ps-4 pe-4">
      <div className="logo">
        <img src="/images/logo.png" alt="logo" />
      </div>
      <nav className="navbar navbar-expand-lg bg-body-tertiary">
        <button
          className="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span className="navbar-toggler-icon"></span>
        </button>
        <div className="collapse navbar-collapse" id="navbarSupportedContent">
          <ul className="navbar-nav me-auto mb-2 mb-lg-0">
            {renderDataMenu()}
          </ul>
        </div>
      </nav>
      <div className="infor d-flex align-items-center">
        <div className="account me-4">
          <Link href={""} prefetch={false}>
            <i className="fa fa-user" aria-hidden="true"></i>
          </Link>
        </div>
        <Link
          prefetch={false}
          href={""}
          onClick={(e) => handleOpenCard()}
          className="shopping-card d-flex align-items-center"
        >
          <p className="me-2 mb-0">$0.00</p>
          <i className="fa fa-shopping-bag" aria-hidden="true"></i>
        </Link>
      </div>


      {isOpen && <Sidebar onHandleOpenSibar={handleOpenCard} />}
    </div>
  );
};

export default Header;
