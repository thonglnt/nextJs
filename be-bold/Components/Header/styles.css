.menu a {
  color: inherit;
}

.menu:hover a {
  color: var(--paleChestnut);
}

.wrapper-hr-tag {
  height: 3px;
}
.hr-tag {
  width: 80%;
  margin: 0 auto;
  display: none;
  color: var(--paleChestnut);
}

.menu:hover .hr-tag {
  display: flex;
  transition: 0.4s;
}

.account a {
  color: inherit;
}

.account a:hover {
  color: inherit;
  transition: 0.4s;
}

.shopping-card {
  color: inherit;
  text-decoration: none;
}

.shopping-card:hover {
  color: inherit;
  transition: 0.4s;
}