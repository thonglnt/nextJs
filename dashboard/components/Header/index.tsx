'use client'

import { memo, useCallback, useState } from 'react'
import {
  Box,
  Divider,
  IconButton,
  Link,
  ListItemIcon,
  MenuItem,
  Tooltip,
  Typography,
  useTheme,
} from '@mui/material'
import SearchIcon from '@mui/icons-material/Search'
import GroupIcon from '@mui/icons-material/Group'
import NotificationsIcon from '@mui/icons-material/Notifications'
import DraftsOutlinedIcon from '@mui/icons-material/DraftsOutlined'
import { PersonAdd, Settings, Logout } from '@mui/icons-material'
import CloseIcon from '@mui/icons-material/Close'
import { useRouter } from 'next/navigation'

// Components
import Badge from '@components/Badge'
import Avatar from '@components/Avatar'
import SearchDialog from '@components/Search'
import MenuComposition from '@components/Menu'

// Constants
import { MOCK_CONTACTS, MOCK_NOTIFICATIONS } from '@constants/mockData'
import { LOCAL_STORAGE_KEY, ROUTE } from '@constants/index'

const Header = () => {
  const theme = useTheme()
  const route = useRouter()

  const [isOpenSearch, setIsOpenSearch] = useState(false)
  const [isOpenContacts, setIsOpenContacts] = useState(false)
  const [isOpenNotifications, setIsOpenNotifications] = useState(false)

  const HandleModalSearch = () => {
    if (isOpenSearch) {
      setIsOpenSearch(false)
    } else {
      setIsOpenSearch(true)
    }
  }

  const HandleMenuContacts = () => {
    if (isOpenContacts) {
      setIsOpenContacts(false)
    } else {
      setIsOpenContacts(true)
    }
  }

  const HandleMenuNotifications = () => {
    if (isOpenNotifications) {
      setIsOpenNotifications(false)
    } else {
      setIsOpenNotifications(true)
    }
  }

  const handleLogout = useCallback(() => {
    localStorage.removeItem(LOCAL_STORAGE_KEY.USER_ID)
    route.push(ROUTE.LOGIN)
  }, [route])

  return (
    <>
      <Box
        width="100%"
        display="flex"
        justifyContent="flex-end"
        gap={2}
        boxShadow="rgb(100 116 139 / 12%) 0px 1px 4px"
      >
        <Tooltip
          title="Search"
          sx={{
            fontSize: '8px',
            paddingTop: '4px',
          }}
          onClick={HandleModalSearch}
        >
          <IconButton size="small" edge="start" color="inherit">
            <SearchIcon sx={{ width: '20px', height: '20px' }} />
          </IconButton>
        </Tooltip>

        <MenuComposition
          element={
            <Tooltip
              title="Contacts"
              sx={{
                fontSize: '8px',
                paddingTop: '4px',
              }}
              onClick={HandleMenuContacts}
            >
              <GroupIcon sx={{ width: '20px', height: '20px' }} />
            </Tooltip>
          }
          open={false}
        >
          <Box padding="24px" width="380px">
            <Box
              display="flex"
              justifyContent="space-between"
              alignItems="center"
              marginBottom="10px"
            >
              <Typography fontWeight={600}>Contacts</Typography>
            </Box>
            {MOCK_CONTACTS.map((item) => (
              <Box
                marginTop="10px"
                display="flex"
                alignItems="center"
                key={item.id}
                justifyContent="space-between"
              >
                <Box display="flex" alignItems="center">
                  <Avatar alt={item.alt} src={item.img} type="normal" />
                  <Typography marginLeft="15px">{item.name}</Typography>
                </Box>
                {item.online ? (
                  <Badge
                    color="primary"
                    overlap="circular"
                    badgeContent=" "
                    variant="dot"
                    content={0}
                  />
                ) : (
                  <Typography>{item.active} ago</Typography>
                )}
              </Box>
            ))}
          </Box>
        </MenuComposition>

        <MenuComposition
          element={
            <Tooltip
              title="Notifications"
              sx={{
                fontSize: '8px',
                paddingTop: '4px',
              }}
              onClick={HandleMenuNotifications}
            >
              <Badge content={4} color="error" invisible={false}>
                <NotificationsIcon color="action" />
              </Badge>
            </Tooltip>
          }
          open={false}
        >
          <Box padding="24px" width="380px">
            <Box
              display="flex"
              justifyContent="space-between"
              alignItems="center"
            >
              <Typography fontWeight={600}>Notifications</Typography>
              <IconButton>
                <DraftsOutlinedIcon fontSize="small" />
              </IconButton>
            </Box>
            {MOCK_NOTIFICATIONS.map((item) => (
              <Box
                marginTop="10px"
                display="flex"
                alignItems="start"
                key={item.id}
              >
                <Avatar alt={item.alt} src={item.img} type="normal" />
                <Box marginLeft="15px">
                  <Typography fontWeight={600} fontSize="14px">
                    {item.content}
                    <Link color={theme.palette.secondary.main} marginLeft="3px">
                      {item.link}
                    </Link>
                  </Typography>
                  <Typography
                    fontSize="12px"
                    color={theme.palette.grey[500]}
                    marginTop="5px"
                  >
                    {item.time}
                  </Typography>
                </Box>
                <IconButton>
                  <CloseIcon fontSize="small" />
                </IconButton>
                <Divider />
              </Box>
            ))}
          </Box>
        </MenuComposition>

        <MenuComposition
          element={
            <Avatar
              type="normal"
              alt="avatar"
              sx={{ width: '20px', height: '20px' }}
            />
          }
          open={false}
        >
          <MenuItem
            sx={{
              gap: '15px',
            }}
          >
            <Avatar type="normal" sx={{ width: '20px', height: '20px' }} />
            <Typography fontSize={14}>Profile</Typography>
          </MenuItem>
          <MenuItem
            sx={{
              gap: '15px',
            }}
          >
            <Avatar type="normal" sx={{ width: '20px', height: '20px' }} />
            <Typography fontSize={14}>My account</Typography>
          </MenuItem>
          <Divider />
          <MenuItem>
            <ListItemIcon>
              <PersonAdd fontSize="small" />
            </ListItemIcon>
            Add another account
          </MenuItem>
          <MenuItem>
            <ListItemIcon>
              <Settings fontSize="small" />
            </ListItemIcon>
            Settings
          </MenuItem>
          <MenuItem onClick={handleLogout}>
            <ListItemIcon>
              <Logout fontSize="small" />
            </ListItemIcon>
            Logout
          </MenuItem>
        </MenuComposition>
      </Box>

      {/* Modal Search */}
      <SearchDialog isOpen={isOpenSearch} onClose={HandleModalSearch} />
    </>
  )
}

export default memo(Header)
