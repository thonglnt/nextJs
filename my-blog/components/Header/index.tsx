'use client'

import { Box, Container, Typography, useTheme } from '@mui/material'

// Constants
import { DEFAULT_HEADER_URL } from '@constants/router'
import { LOGO_TITLE, NAVIGATE_ITEM } from '@constants/__mock__/mockData'

// Components
import Logo from '@components/Logo'
import Navigation from '@components/Navigation'
import Breadcrumbs, { LinkItem } from '@components/Breadcrumbs'

export interface IHeader {
  pageTitle: string
  breadCrumbs: LinkItem[]
}

const Header = ({ pageTitle = 'Title', breadCrumbs = [] }: IHeader) => {
  const theme = useTheme()

  return (
    <Box component="header" marginBottom="50px">
      <Box paddingTop="20px" bgcolor={theme.palette.grey[100]}>
        <Container
          maxWidth="lg"
          sx={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'center',
          }}
        >
          <Logo
            link={DEFAULT_HEADER_URL.HOME.URL}
            marginBottom="20px"
            letterSpacing="2px"
          >
            {LOGO_TITLE}
          </Logo>
          <Navigation navItems={NAVIGATE_ITEM} />
        </Container>
      </Box>
      {breadCrumbs.length > 0 && (
        <Box
          paddingTop="20px"
          bgcolor={theme.palette.grey[100]}
          sx={{
            paddingTop: {
              xs: '50px',
              md: '100px',
            },
          }}
        >
          <Container maxWidth="lg">
            <Breadcrumbs links={breadCrumbs} />
            <Typography
              data-testid="page-title"
              variant="h2"
              letterSpacing="2px"
              fontWeight={800}
              color={theme.palette.grey[400]}
              textTransform="uppercase"
              sx={{
                fontSize: {
                  xs: '25px',
                  md: '48px',
                },
                marginTop: {
                  xs: '20px',
                  md: '40px',
                },
              }}
            >
              {pageTitle}
            </Typography>
          </Container>
        </Box>
      )}
    </Box>
  )
}

export default Header
