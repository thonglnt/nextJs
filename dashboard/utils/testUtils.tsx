import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'

// Themes
import theme from '@self-types/Themes.types'

export const customRender = (children: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
}

export * from '@testing-library/react'
export { customRender as render }
