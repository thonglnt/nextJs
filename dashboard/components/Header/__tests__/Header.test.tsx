// Utils
import { fireEvent, render, screen } from '@utils/testUtils'
import { useRouter } from 'next/navigation'

// Components
import Header from '..'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

describe('Header component', () => {
  it('should render Header component', () => {
    const { container } = render(<Header />)

    expect(container).toMatchSnapshot()
  })

  it('should render popup search after click button search and close after click button close ', () => {
    render(<Header />)

    const buttonSearch = screen.getByTestId('SearchIcon')

    fireEvent.click(buttonSearch)

    const modalSearch = screen.getByTestId('modalSearch')

    expect(modalSearch).toBeInTheDocument()

    const closeButton = screen.getByTestId('CloseIcon')

    fireEvent.click(closeButton)
  })

  it('should render popup contacts after click button contact and close after click again button contact ', () => {
    render(<Header />)

    const buttonContact = screen.getByTestId('GroupIcon')

    fireEvent.click(buttonContact)

    const popupContact = screen.getByText(/contacts/i)

    expect(popupContact).toBeInTheDocument()

    fireEvent.click(buttonContact)
  })

  it('should render popup notification after click button notification and close after click again button notification ', () => {
    render(<Header />)

    const buttonNotifications = screen.getByTestId('NotificationsIcon')

    fireEvent.click(buttonNotifications)

    const popupNotification = screen.getByRole('menu')

    expect(popupNotification).toBeInTheDocument()

    fireEvent.click(buttonNotifications)
  })

  it('should render popup user after click button user and logout user after click logout ', () => {
    const push = jest.fn()

    ;(useRouter as jest.Mock).mockImplementation(() => ({
      push,
    }))

    render(<Header />)

    const buttonUsers = screen.getByTestId('PersonIcon')

    fireEvent.click(buttonUsers)

    const popupUser = screen.getByRole('menu')

    expect(popupUser).toBeInTheDocument()

    const buttonLogout = screen.getByRole('menuitem', {
      name: /logout/i,
    })

    fireEvent.click(buttonLogout)
    expect(push).toHaveBeenCalledWith('/login')
  })
})
