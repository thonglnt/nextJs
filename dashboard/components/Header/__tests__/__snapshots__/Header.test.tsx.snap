// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header component 1`] = `
<div>
  <div
    class="MuiBox-root css-8zh55s"
  >
    <button
      aria-label="Search"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeSmall css-1bfgnep-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xhbeow-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        aria-label="Contacts"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xhbeow-MuiSvgIcon-root"
        data-mui-internal-clone-element="true"
        data-testid="GroupIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5c-1.66 0-3 1.34-3 3s1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5C6.34 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <span
        aria-label="Notifications"
        class="MuiBadge-root MuiBadge-root css-dhntzm-MuiBadge-root"
        data-mui-internal-clone-element="true"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorAction MuiSvgIcon-fontSizeMedium css-11pbyhm-MuiSvgIcon-root"
          data-testid="NotificationsIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
          />
        </svg>
        <span
          class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightCircular MuiBadge-overlapCircular MuiBadge-colorError MuiBadge-badge css-10tn45-MuiBadge-badge"
        >
          4
        </span>
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-x9hid5-MuiAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
